src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2004 @@
+    } else if(index >= len1) {
+      val io = index - len1
+      prefix2(io >>> BITS)(io & MASK)
+    } else prefix1(index)
+  }.asInstanceOf[A]
+
+  override def updated[B >: A](index: Int, elem: B): NVector[B] = {
+    if(index < 0 || index >= length) throw ioob(index)
+    if(index >= len123) {
+      val io = index - len123
+      val i4 = io >>> BITS3
+      val i3 = (io >>> BITS2) & MASK
+      val i2 = (io >>> BITS) & MASK
+      val i1 = io & MASK
+      if     (i4 < data4.length  ) copy(data4   = copyUpdate(data4, i4, i3, i2, i1, elem))
+      else if(i3 < suffix3.length) copy(suffix3 = copyUpdate(suffix3, i3, i2, i1, elem))
+      else if(i2 < suffix2.length) copy(suffix2 = copyUpdate(suffix2, i2, i1, elem))
+      else                         copy(suffix1 = copyUpdate(suffix1, i1, elem))
+    } else if(index >= len12) {
+      val io = index - len12
+      copy(prefix3 = copyUpdate(prefix3, io >>> BITS2, (io >>> BITS) & MASK, io & MASK, elem))
+    } else if(index >= len1) {
+      val io = index - len1
+      copy(prefix2 = copyUpdate(prefix2, io >>> BITS, io & MASK, elem))
+    } else {
+      copy(prefix1 = copyUpdate(prefix1, index, elem))
+    }
+  }
+
+  override def appended[B >: A](elem: B): NVector[B] = {
+    if     (suffix1.length < WIDTH  ) copy(suffix1 = copyAppend(suffix1, elem.asInstanceOf[AnyRef]), length = length+1)
+    else if(suffix2.length < WIDTH-1) copy(suffix2 = copyAppend(suffix2, suffix1), suffix1 = wrap1(elem), length = length+1)
+    else if(suffix3.length < WIDTH-1) copy(suffix3 = copyAppend(suffix3, copyAppend(suffix2, suffix1)), suffix2 = empty2, suffix1 = wrap1(elem), length = length+1)
+    else if(data4.length   < WIDTH-2) copy(data4   = copyAppend(data4, copyAppend(suffix3, copyAppend(suffix2, suffix1))), suffix3 = empty3, suffix2 = empty2, suffix1 = wrap1(elem), length = length+1)
+    else new NVector5(prefix1, len1, prefix2, len12, prefix3, len123, data4, (WIDTH-2)*WIDTH3 + len123, empty5, wrap4(copyAppend(suffix3, copyAppend(suffix2, suffix1))), empty3, empty2, wrap1(elem), length+1)
+  }
+
+  override def prepended[B >: A](elem: B): NVector[B] = {
+    if     (len1         < WIDTH  ) copy(copyPrepend1(elem, prefix1), len1+1, len12 = len12+1, len123 = len123+1, length = length+1)
+    else if(len12        < WIDTH2 ) copy(wrap1(elem), 1, copyPrepend(prefix1, prefix2), len12+1, len123 = len123+1, length = length+1)
+    else if(len123       < WIDTH3 ) copy(wrap1(elem), 1, empty2, 1, copyPrepend(copyPrepend(prefix1, prefix2), prefix3), len123+1, length = length+1)
+    else if(data4.length < WIDTH-2) copy(wrap1(elem), 1, empty2, 1, empty3, 1, copyPrepend(copyPrepend(copyPrepend(prefix1, prefix2), prefix3), data4), length = length+1)
+    else new NVector5(wrap1(elem), 1, empty2, 1, empty3, 1, wrap4(copyPrepend(copyPrepend(prefix1, prefix2), prefix3)), len123+1, empty5, data4, suffix3, suffix2, suffix1, length+1)
+  }
+
+  override def foreach[U](f: A => U): Unit = {
```

Comment:
In practice, is this faster than:
@tailrec def foreachN(n: Int): Unit =
```suggestion
  override def foreach[U](f: A => U): Unit = {
```
  if (n < 7) {
    foreachElem(vectorSlice(n), f)
    foreachN(n + 1)
  }

foreachN(0)

Reply:
If you implement this method like this as a final method , then allow for a `slices`-method override, then you could have a single implementation of foreach.

Reply:
This is generally counter-productive. You want to keep virtual method calls with megamorphic call sites at the outer layer (if necessary at all) and then keep eveything nicely monomorphic inside. I haven't tested this exact version but I did try a generic `foreach` that uses vector slices and it was slower.

Reply:
Well, that's what I get for saying "generally counter-productive". I benchmarked it (using the existing `foreachRec` helper method) and it's actually faster.

Reply:
@szeiger You're welcome (Old Klangian Perf Hack)